src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,136 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int BASIC_LENGTH = 10;
+    private T[] list;
+    private int size;
+
+    public ArrayList() {
+        list = (T[]) new Object[BASIC_LENGTH];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == list.length) {
+            addSpase();
+        }
+        list[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (!(index > size || index < 0)) {
+            if (index == size) {
+                /* If index of element fall into the end of array */
+                add(value);
+            } else {
+                if (list.length == size) {
+                    /* If array is overflowing, we need increase the array */
+                    addSpase();
+                    addByIndex(value, index);
+                } else {
+                    addByIndex(value, index);
+                }
+            }
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index of list");
+        }
+    }
+
+    private void addSpase() {
+        T[] array = (T[]) new Object[list.length + (list.length >> 1)];
+        System.arraycopy(list, 0, array, 0, list.length);
+        list = array;
+    }
 
+    private void addByIndex(T value, int index) {
+        T[] array = (T[]) new Object[list.length];
+        System.arraycopy(list,0,array,0,index);
+        array[index] = value;
```

Comment:
Redundant object. You only use this if we don't have enough space and need to increase it.